### DISCUSSION TITLE
ParaView: Mixed definition of unstructured zone by elements and by faces is not valid

### DISCUSSION THREAD

**Discussion by dbgarlisch** [CGNS/CGNS#720] [CGNStalk]
I found this [1 year old CGNS discussion][1] and a similar [2 year old ParaView discussion][2].

As best I can tell, CGNS/SIDS does NOT prohibit mixed element/face zones.

**Am I correct in saying that it is the ParaView CGNS reader that is imposing this limitation?**

I don't know of any reason why a reader shouldn't be able to handle mixed element/face zones. It might be a little more work/bookkeeping; but doable.

I have generated mixed element/face zones meshes that load into CGNSview and pass CGNScheck (v3.31) without objection.

I know polyhedron/polygon encoding changes in v4. **Is this a restriction new to v4?**

I am looking for *official* guidance on what a valid CGNS export should be for a mesh volume zone that contains both canonical and polyhedron elements...

**Do I split out the polyhedron and canonical cells into different zones with the same VC?**
**Do I write all canonical cells as polyhedron?**
**Other?**

**Or is this a practical issue?** Since many/most/all readers will most likely balk on a mixed element/face zone, I should play it safe and export a _uniform_ file?

Whatever the official answer, the SIDS and related docs should be amended to make this clear.

[1]: https://github.com/CGNS/CGNS/discussions/345
[2]: https://discourse.paraview.org/t/error-reading-cgns-file-with-arbitrary-polyhedral-elements/6286

**Comment by brtnfld** [comment]
FYI, VisIt https://wci.llnl.gov/simulation/computer-codes/visit if you want to see if it can read the file.

**Comment by dbgarlisch** [comment]
The ParaView messages for [HemisphereInBox-Poly.zip][1]:

> ERROR: In vtkCGNSReader.cxx, line 2471
> vtkCGNSReader (000001F4A6F9FE10): Mixed definition of unstructured zone by elements and by faces is not valid.
> 
> ERROR: In vtkCGNSReader.cxx, line 4713
> vtkCGNSReader (000001F4A6F9FE10): Error Reading file
> 
> ERROR: In vtkExecutive.cxx, line 741
> vtkPVCompositeDataPipeline (000001F4A694FBA0): Algorithm vtkCGNSFileSeriesReader (000001F4A6950920) returned failure for request: vtkInformation (000001F4B8765A30)
>   Debug: Off
>   Modified Time: 457447
>   Reference Count: 1
>   Registered Events: (none)
>   Request: REQUEST_DATA
>   FORWARD_DIRECTION: 0
>   ALGORITHM_AFTER_FORWARD: 1
>   FROM_OUTPUT_PORT: 0

The TET (red) and PYRAMID (yellow) cells in the transition regions between the HEX (blue) cells are converted to polyhedra when exported to the CGNS file. The hex cells with faces touching the transition pyramids are also converted to polyhedra. The resulting mesh is fully point matched (no hanging nodes).

![image](https://github.com/CGNS/CGNS/assets/4324473/12ded4d6-fe28-48ab-a541-794e6964fdb3)


[1]: https://github.com/CGNS/CGNS/files/12016525/HemisphereInBox-Poly.zip

**Comment by dbgarlisch** [comment]
I misspoke earlier. Due to recent Fluent export changes, the ParaView `Extract Cells by Type` filter is no longer crashing.

However, after importing a [Fluent cas version of the mesh][1], the polyhedron cells do not display when the filter is activated.

![image](https://github.com/CGNS/CGNS/assets/4324473/058a1074-c239-45cf-8df9-2aba588e70e8)
![image](https://github.com/CGNS/CGNS/assets/4324473/ac5ef2eb-7434-44d7-b22b-2fa6345e4502)


[1]: https://github.com/CGNS/CGNS/files/12017225/HemisphereInBox-Poly-cas.zip

**Comment by MicK7** [comment]
If one stores in 3 separate Elements_t sections the HEX,  the TET and the PYRAMID and in an other section the POLYHEDRON under the same Zone. It should be a valid CGNS Zone that ParaView should handle. What is not supported is having the HEX, TET and PYRAMID in the same MIXED section with a POLYHEDRAL section.
